app/src/main/java/com/keylesspalace/tusky/components/timeline/TimelineFragment.kt

```diff
@@ -293,23 +170,277 @@ class TimelineFragment :
+                    }
+
+                    // Refresh adapter on mutes and blocks
+                    launch {
+                        viewModel.uiSuccess.collectLatest {
+                            when (it) {
+                                is UiSuccess.Block,
+                                is UiSuccess.Mute,
+                                is UiSuccess.MuteConversation ->
+                                    adapter.refresh()
+
+                                is UiSuccess.StatusSent -> handleStatusSentOrEdit(it.status)
+                                is UiSuccess.StatusEdited -> handleStatusSentOrEdit(it.status)
+
+                                else -> { /* nothing to do */ }
+                            }
+                        }
                     }
-                    is StatusEditedEvent -> {
-                        handleStatusComposeEvent(event.status)
+                }
+
+                // Update status display from statusDisplayOptions. If the new options request
+                // relative time display collect the flow to periodically re-bind the UI.
+                launch {
+                    viewModel.statusDisplayOptions
+                        .collectLatest {
+                            adapter.statusDisplayOptions = it
+                            layoutManager.findFirstVisibleItemPosition().let { first ->
+                                first == RecyclerView.NO_POSITION && return@let
+                                val count = layoutManager.findLastVisibleItemPosition() - first
+                                adapter.notifyItemRangeChanged(
+                                    first,
+                                    count,
+                                    null
+                                )
+                            }
+
+                            if (!it.useAbsoluteTime) {
+                                updateTimestampFlow.collect()
+                            }
+                        }
+                }
+
+                // Scroll the list down if a refresh has completely finished. A refresh is
+                // finished when both the initial refresh is complete and any prepends have
+                // finished (so that DiffUtil has had a chance to process the data). See
+                // https://github.com/googlecodelabs/android-paging/issues/149
+                launch {
+                    var old: CombinedLoadStates? = null
+                    var refreshComplete = false
+
+                    if (isSwipeToRefreshEnabled) {
+                        adapter.loadStateFlow
+                            .collect {
+                                if (old == null) {
+                                    Log.d("loadState", "No previous loadState")
+                                    old = it
+                                    return@collect
+                                }
+
+                                if (old?.refresh != it.refresh) {
```

Comment:
This section here can be cleaned up, or moved to a function to clean up. Trying to figure out the specific state transitions to get the UI updated is tricky.